eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala

```diff
@@ -1460,52 +1452,28 @@ object Helpers {
       val localCommit = d.commitments.latest.localCommit
       val remoteCommit = d.commitments.latest.remoteCommit
       val nextRemoteCommit_opt = d.commitments.latest.nextRemoteCommit_opt.map(_.commit)
+      // NB: from the p.o.v of remote, their incoming htlcs are our outgoing htlcs.
+      val outgoingHtlcs = localCommit.spec.htlcs.collect(outgoing) ++ (remoteCommit.spec.htlcs ++ nextRemoteCommit_opt.map(_.spec.htlcs).getOrElse(Set.empty)).collect(incoming)
       if (localCommit.commitTxAndRemoteSig.commitTx.tx.txid == tx.txid) {
-        // our commit got confirmed, so any htlc that is in their commitment but not in ours will never reach the chain
-        val htlcsInRemoteCommit = remoteCommit.spec.htlcs ++ nextRemoteCommit_opt.map(_.spec.htlcs).getOrElse(Set.empty)
-        // NB: from the p.o.v of remote, their incoming htlcs are our outgoing htlcs
-        htlcsInRemoteCommit.collect(incoming) -- localCommit.spec.htlcs.collect(outgoing)
+        // Our commit got confirmed: any htlc that is *not* in our commit will never reach the chain.
+        outgoingHtlcs -- localCommit.spec.htlcs.collect(outgoing)
       } else if (d.revokedCommitPublished.map(_.commitTx.txid).contains(tx.txid)) {
-        // a revoked commitment got confirmed: we will claim its outputs, but we also need to fail htlcs that are pending in the latest commitment:
-        //  - outgoing htlcs that are in the local commitment but not in remote/nextRemote have already been fulfilled/failed so we don't care about them
-        //  - outgoing htlcs that are in the remote/nextRemote commitment may not really be overridden, but since we are going to claim their output as a
-        //    punishment we will never get the preimage and may as well consider them failed in the context of relaying htlcs
-        nextRemoteCommit_opt.getOrElse(remoteCommit).spec.htlcs.collect(incoming)
+        // A revoked commitment got confirmed: we will claim its outputs, but we also need to resolve upstream htlcs.
+        // We consider *all* outgoing htlcs failed: our peer may reveal the preimage with an HTLC-success transaction,
+        // but it's more likely that our penalty transaction will confirm first. In any case, since we will get those
+        // funds back on-chain, it's as if the outgoing htlc had failed and should thus be failed back upstream.
```

Comment:
> In any case, since we will get those funds back on-chain, it's as if the outgoing htlc had failed and should thus be failed back upstream.

I don't understand that comment. An htlc can be failed downstream and fulfilled upstream. I do understand why it makes sense to fail all outgoing htlcs in this case.